src/Tools/dotnet-monitor/Egress/S3/MultiPartUploadStream.cs

```diff
@@ -47,13 +59,43 @@ public override async Task FlushAsync(CancellationToken cancellationToken)
 
     public async Task FinalizeAsync(CancellationToken cancellationToken)
     {
+        _finalize = true;
+
+        if (_writeSynchronousArtifacts != null)
+            await _writeSynchronousArtifacts;
+        
         if (Closed)
             throw new ObjectDisposedException(nameof(MultiPartUploadStream));
         if (_offset == 0)
             return;
         await DoWriteAsync(true, cancellationToken);
     }
 
+    public async Task WriteAsyncLoop(CancellationToken cancellationToken)
+    {
+        while (!_finalize || _syncTempBuffer.Count > 0)
+        {
+            if (Closed)
+                throw new ObjectDisposedException(nameof(MultiPartUploadStream));
+
+            await _semaphore.WaitAsync(cancellationToken);
+            try
+            {
+                if (_syncTempBuffer.Count > 0)
+                {
+                    await WriteAsync(_syncTempBuffer.ToArray(), _finalize, cancellationToken);
+                    _syncTempBuffer = new();
```

Comment:
Clear the buffer instead of creating a new one to avoid extra allocations.

Reply:
I had that initially but changed it to avoid situations where the `_syncTempBuffer` expands to a large size due to a brief spike in logs and then remains unnecessarily large after that point. I can change this back to clearing though if you'd prefer.

Reply:
You can clear it and then set Capacity to some reasonable value so we avoid unnecessary allocations of the `List<byte>` itself and unnecessary reallocation of the array within it each time `AddRange` is called.